oximeter/db/tests/integration_test.rs

```diff
@@ -0,0 +1,380 @@
+    );
+    client3.insert_samples(&samples).await.expect("failed to insert samples");
+
+    // Ensure that both replica 3 and replica 2 have the samples
+    wait_for_num_points(&client3, samples.len() * 2)
+        .await
+        .expect("failed to get samples from client3");
+    wait_for_num_points(&client2, samples.len() * 2)
+        .await
+        .expect("failed to get samples from client2");
+
+    // Restart node 1 and wait for the data to replicate
+    deployment.start_server(1).expect("failed to restart clickhouse server 1");
+    wait_for_ping(&client1).await.expect("failed to ping server 1");
+    wait_for_num_points(&client1, samples.len() * 2)
+        .await
+        .expect("failed to get samples from client1");
+
+    // Remove a keeper node
+    deployment.remove_keeper(2).expect("failed to remove keeper");
+
+    // Querying from any node should still work after a keeper is removed
+    wait_for_num_points(&client1, samples.len() * 2)
+        .await
+        .expect("failed to get samples from client1");
+
+    // We still have a quorum (2 of 3 keepers), so we should be able to insert
+    let samples = test_util::generate_test_samples(
+        input.n_projects,
+        input.n_instances,
+        input.n_cpus,
+        input.n_samples,
+    );
+    client3.insert_samples(&samples).await.expect("failed to insert samples");
+    wait_for_num_points(&client2, samples.len() * 3)
+        .await
+        .expect("failed to get samples from client1");
+
+    // Stop another keeper
+    deployment.stop_keeper(1).expect("failed to stop keeper");
+
+    // We should still be able to query
+    wait_for_num_points(&client3, samples.len() * 3)
+        .await
+        .expect("failed to get samples from client1");
+
+    println!("Attempting to insert samples without keeper quorum");
+    let samples = test_util::generate_test_samples(
+        input.n_projects,
+        input.n_instances,
+        input.n_cpus,
+        input.n_samples,
+    );
+    // We have lost quorum and should not be able to insert
+    client1
+        .insert_samples(&samples)
+        .await
+        .expect_err("insert succeeded without keeper quorum");
+
+    // Bringing the keeper back up should allow us to insert again
+    deployment.start_keeper(1).expect("failed to restart keeper");
+    wait_for_keepers(&deployment, vec![1, 3])
+        .await
+        .expect("failed to sync keepers");
+    let samples = test_util::generate_test_samples(
+        input.n_projects,
+        input.n_instances,
+        input.n_cpus,
+        input.n_samples,
+    );
+    client1.insert_samples(&samples).await.expect("failed to insert samples");
+    wait_for_num_points(&client2, samples.len() * 4)
+        .await
+        .expect("failed to get samples from client1");
+
+    // Add a new keeper to restore fault tolerance and try again
+    deployment.add_keeper().expect("Failed to add keeeper");
+    wait_for_keepers(&deployment, vec![1, 3, 4])
```

Comment:
This is more of a `clickward` question, but we might want to make these IDs a newtype to avoid conflating them with an _index_ of the node.

Reply:
Great idea! I'll go ahead and make that change in clickward and use it here. I was definitely being a tad lazy by using u64s here.